CDCgov/NEDSS-Modernization
# Reorder page content

## Description

An API needs to exist to allow users to reorder the content of a Pagebuilder page. This API should allow moving 
1. Tabs
2. Sections
3. Subsections
4. Questions

Given the following page content for page 1234:
|Element Type | Id | 
|----|-----|
|Tab|1|
|Section | 2|
|Subsection|3|
|Question|4|
|Question|5|
|Tab|6|
|Section|7|
|Subsection|8|

Moving the first question into Subsection with Id 8

Would be 
PUT `/api/v1/pages/1234/component/4/after/8`

Or moving the question with Id 4 to after the question with Id 5

PUT `/api/v1/pages/1234/component/4/after/5`


## Tickets

* [CNFT2-718](https://cdc-nbs.atlassian.net/browse/CNFT2-718)

## Steps to verify

1. Log in to NBS as a user with access to page builder (superuser)
1. View a page that has a draft
1. Using swagger, send a request to move an element. you will need to access the database to get the correct Ids
1. Refresh the Draft view in NBS and verify element was moved


Query to see the content of a page
```sql
SELECT
	order_nbr,
	NBS_ui_component.type_cd_desc,
	WA_UI_metadata.wa_ui_metadata_uid,
	question_label,	
	NBS_ui_component.nbs_ui_component_uid
FROM
	WA_UI_metadata
JOIN NBS_ui_component ON NBS_ui_component.nbs_ui_component_uid = WA_UI_metadata.nbs_ui_component_uid
WHERE	
	 wa_template_uid = 1000377 
	 AND order_nbr > 0
ORDER BY
	order_nbr ASC;
```

[CNFT2-718]: https://cdc-nbs.atlassian.net/browse/CNFT2-718?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Sample results from query
<img width="933" alt="image" src="https://github.com/CDCgov/NEDSS-Modernization/assets/109251240/aca3a7e1-bb82-42e9-ad6e-9d1144d0b30a">


If you wanted to Move the `Information As of Date` question to after the `Comments` question then you would send the following request:


PUT `/api/v1/pages/1000377/components/1119289/after/1119290`
